[msan] Implement support for Arm NEON vst{2,3,4} instructions

thurstond opened the pull request:
This adds support for vst{2,3,4}, which are not correctly handled by handleUnknownIntrinsic/handleVector{Load,Store}Intrinsic.

This patch also updates the tests introduced in https://github.com/llvm/llvm-project/pull/98247 and https://github.com/llvm/llvm-project/pull/99555

thurstond committed:
[msan] Implement support for Arm NEON vst{2,3,4} instructions

This adds support for vst{2,3,4}, which are not correctly handled by
handleUnknownIntrinsic/handleVector{Load,Store}Intrinsic.

thurstond committed:
Update code and test

thurstond committed:
Update instrumentation and test

thurstond committed:
Simplify code and reformat

thurstond committed:
Remove unnecessary braces

thurstond committed:
Skip last two operands

thurstond committed:
Origin tracking

thurstond committed:
Add preconditions for interleave functions

thurstond committed:
Comment

thurstond committed:
Wording

vitalybuka commented on llvm/test/Instrumentation/MemorySanitizer/AArch64/neon_vst.ll:
```diff
@@ -0,0 +1,2825 @@
+; CHECK-NEXT:    [[TMP9:%.*]] = icmp ne i64 [[TMP1]], 0
+; CHECK-NEXT:    [[TMP10:%.*]] = select i1 [[TMP9]], i32 [[TMP2]], i32 0
+; CHECK-NEXT:    [[_MSCMP:%.*]] = icmp ne i64 [[TMP1]], 0
+; CHECK-NEXT:    br i1 [[_MSCMP]], label [[TMP11:%.*]], label [[TMP12:%.*]], !prof [[PROF0]]
```
can we precommit the test with the current state of instrumentation?

thurstond replied:
https://github.com/llvm/llvm-project/pull/98247

vitalybuka replied:
on undef msan can assume all 1s and it can skip some instructions

thurstond replied:
Changed to non-track-origins in https://github.com/llvm/llvm-project/pull/99360/commits/cad5cf0dfeb39da75b96195de4b73e15963180c1

vitalybuka commented on llvm/test/Instrumentation/MemorySanitizer/AArch64/neon_vst.ll:
```diff
@@ -0,0 +1,2825 @@
+; CHECK-NEXT:    call void @llvm.aarch64.neon.st4.v2i64.p0(<2 x i64> undef, <2 x i64> undef, <2 x i64> [[C]], <2 x i64> undef, ptr [[P]])
+; CHECK-NEXT:    ret void
+;
+  call void @llvm.aarch64.neon.st4.v2i64.p0(<2 x i64> undef, <2 x i64> undef, <2 x i64> %C, <2 x i64> undef, ptr %P)
```
why do we use "undef" in test, I'd expect %A here?

vitalybuka replied:
on undef msan can assume all 1s and it can skip some instructions

thurstond replied:
There are multiple versions of the test, some with undefs, and some without

e.g.,
void @st4_2d: fully defined
void @st4_2d_undefA: A is undefined
void @st4_2d_undefABD: A, B, D are undefined
...

thurstond replied:
> on undef msan can assume all 1s and it can skip some instructions

That's why there are undef versions of the test - it's to show/check that MSan can work optimally.

vitalybuka replied:
I see, I didn't notice that, they kind of redundant, because this is generic part of msan which should short cut undefs, but I don't mind to have new tests her

thurstond committed:
Remove track-origins from test

vitalybuka commented on llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:
```diff
@@ -3742,6 +3751,124 @@ struct MemorySanitizerVisitor : public InstVisitor<MemorySanitizerVisitor> {
+
+    if (MS.TrackOrigins) {
+      // We don't use the last two operands to compute the origin, because:
+      // - the last operand is the callee
```
We should operate on `I.arg_size()` and `getArgOperand` the we will have no callee here

thurstond replied:
Done

thurstond committed:
Use arg_size() instead of (getNumOperands() - 1)

vitalybuka commented on llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:
```diff
@@ -2483,13 +2484,21 @@ struct MemorySanitizerVisitor : public InstVisitor<MemorySanitizerVisitor> {
-  void setOriginForNaryOp(Instruction &I) {
+  ///
+  /// Optionally skips n trailing operands.
+  void setOriginForNaryOp(Instruction &I, unsigned int skipLastOperands = 0) {
```
for this particular case skipping callee is unnecessary, it will be skipped in .Add()
as `getShadow` for that operand should be null.

vitalybuka replied:
I prefer instead of extending this function we just to do 

```
    OriginCombiner OC(this, IRB);
    for (args)
      OC.Add(I.getOperand(i));
```

inside of

thurstond replied:
> I prefer instead of extending this function we just to do
> 
> ```
>     OriginCombiner OC(this, IRB);
>     for (args)
>       OC.Add(I.getOperand(i));
> ```
> 
> inside of handleNEONVectorStoreIntrinsic

Done in https://github.com/llvm/llvm-project/pull/99360/commits/1970568ab31879094666b20d6f09b39682c20864

vitalybuka commented on llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:
```diff
@@ -3742,6 +3751,124 @@ struct MemorySanitizerVisitor : public InstVisitor<MemorySanitizerVisitor> {
+    Value *interleavedShadow = interleaveShadow(IRB, I);
+
+    // Call arguments only
+    int numArgOperands = I.getNumOperands() - 1;
```
I believe we add a check that result pointer itself is initialized

thurstond replied:
I've moved the pointer check earlier in https://github.com/llvm/llvm-project/pull/99360/commits/1cebde812e77dadbe38a39cd21878a79605e409c

vitalybuka commented on llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:
```diff
@@ -3742,6 +3751,124 @@ struct MemorySanitizerVisitor : public InstVisitor<MemorySanitizerVisitor> {
+
+  /// Calculates the shadow for interleaving 2, 3 or 4 vectors
+  /// (e.g., for Arm NEON vector store).
+  Value *interleaveShadow(IRBuilder<> &IRB, IntrinsicInst &I) {
```
Another version of the instrumentation could be just clone instruction and use shadow values as arguments. It's probably OK here.

thurstond replied:
Good idea! Done in https://github.com/llvm/llvm-project/pull/99360/commits/41d3a53f770bd78cc5ab7325ac1f8f62437dd73c

thurstond committed:
Revert setOriginForNaryOp

vitalybuka commented on llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:
```diff
@@ -3865,6 +3866,125 @@ struct MemorySanitizerVisitor : public InstVisitor<MemorySanitizerVisitor> {
+    if (MS.TrackOrigins) {
+      OriginCombiner OC(this, IRB);
+      for (int i = 0; i < numArgOperands - 1; i++)
+        OC.Add(I.getOperand(i));
```
Please iterate over args, not operands

thurstond replied:
Fixed in https://github.com/llvm/llvm-project/pull/99360/commits/7a770ddec98c9dac7fac2e79ca4903272625cac7

thurstond committed:
Move pointer check earlier

thurstond committed:
Change from getOperand to getArgOperand

thurstond committed:
Formatting

thurstond committed:
Clone instruction, as suggested by Vitaly

thurstond committed:
Formatting

vitalybuka commented on llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:
```diff
@@ -3865,6 +3866,42 @@ struct MemorySanitizerVisitor : public InstVisitor<MemorySanitizerVisitor> {
+      insertShadowCheck(Addr, &I);
+
+    IntrinsicInst *ShadowI = cast<IntrinsicInst>(I.clone());
+    for (int i = 0; i < numArgOperands - 1; i++) {
```
no {} in one liners

thurstond replied:
I ended up making it a two-liner, to assert that the arg operand is a fixed vector type

vitalybuka commented on llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:
```diff
@@ -3865,6 +3866,42 @@ struct MemorySanitizerVisitor : public InstVisitor<MemorySanitizerVisitor> {
+      OriginCombiner OC(this, IRB);
+      for (int i = 0; i < numArgOperands - 1; i++)
+        OC.Add(I.getArgOperand(i));
+      OC.Done(&I);
```
OriginCombiner uses combined origin as origin of instruction return value

here `void Done(Instruction *I) {` dose not work
we likely need another `DoneAndStore` for ptr like cases

vitalybuka replied:
you may want to fork a single trivial case into another test file with origin tracking

thurstond replied:
Done in https://github.com/llvm/llvm-project/pull/99360/files

thurstond committed:
Fix origin painting (contains debugging output)

thurstond committed:
Update origin

thurstond committed:
Formatting

thurstond committed:
Formatting

thurstond committed:
Add comment about origin tracking imprecision

vitalybuka commented on llvm/lib/Transforms/Instrumentation/MemorySanitizer.cpp:
```diff
@@ -2498,6 +2499,15 @@ struct MemorySanitizerVisitor : public InstVisitor<MemorySanitizerVisitor> {
+    void DoneAndStoreOrigin(TypeSize TS, Value *OriginPtr) {
+      if (MSV->MS.TrackOrigins) {
+        assert(Origin);
+        MSV->paintOrigin(IRB, Origin, OriginPtr, TS, Align(1));
```
Origin aligment must be 4+

`const Align OriginAlignment = std::max(kMinOriginAlignment, Alignment);`
and I suspect this instruction guaranty some `Alignment`?

vitalybuka replied:
I guess LGTM, if we don't know Ptr Alignment

thurstond replied:
Updated to kMinOriginAlignment in https://github.com/llvm/llvm-project/pull/99360/commits/2d0abae869d0c565dc6230ec8adbaf5067edfc03

thurstond replied:
Changed back to 1-byte alignment in https://github.com/llvm/llvm-project/pull/99360/commits/af86d86e1d00f09c4e8c8f425a5bca6522636820 per offline discussion

thurstond committed:
Paint origin with kMinOriginAlignment

thurstond committed:
Relax alignment back to 1 byte because it is not guaranteed to be more

"element size has an impact on pointer alignment. Alignment to the element size generally gives better performance, and it might be a requirement of your target operating system. For example, when loading 32-bit elements, align the address of the first element to at least 32-bits."
 -- https://developer.arm.com/documentation/102159/0400/Load-and-store---data-structures

thurstond committed:
Update test with relaxed alignment

vitalybuka committed:
OriginPtr is aligned by getShadowOriginPtr

vitalybuka committed:
Update neon_vst_origins.ll
